File: android/app/src/main/java/com/oppzippy/openscq30/ui/devicesettings/EqualizerProfile.kt
package com.oppzippy.openscq30.ui.devicesettings

import android.util.Log
import com.oppzippy.openscq30.lib.EqualizerBandOffsets
import com.oppzippy.openscq30.lib.EqualizerConfiguration
import com.oppzippy.openscq30.lib.PresetEqualizerProfile

enum class EqualizerProfile(val presetProfile: PresetEqualizerProfile?) {
    SoundcoreSignature(PresetEqualizerProfile.SoundcoreSignature),
    Acoustic(PresetEqualizerProfile.Acoustic),
    BassBooster(PresetEqualizerProfile.BassBooster),
    BassReducer(PresetEqualizerProfile.BassReducer),
    Classical(PresetEqualizerProfile.Classical),
    Podcast(PresetEqualizerProfile.Podcast),
    Dance(PresetEqualizerProfile.Dance),
    Deep(PresetEqualizerProfile.Deep),
    Electronic(PresetEqualizerProfile.Electronic),
    Flat(PresetEqualizerProfile.Flat),
    HipHop(PresetEqualizerProfile.HipHop),
    Jazz(PresetEqualizerProfile.Jazz),
    Latin(PresetEqualizerProfile.Latin),
    Lounge(PresetEqualizerProfile.Lounge),
    Piano(PresetEqualizerProfile.Piano),
    Pop(PresetEqualizerProfile.Pop),
    RnB(PresetEqualizerProfile.RnB),
    Rock(PresetEqualizerProfile.Rock),
    SmallSpeakers(PresetEqualizerProfile.SmallSpeakers),
    SpokenWord(PresetEqualizerProfile.SpokenWord),
    TrebleBooster(PresetEqualizerProfile.TrebleBooster),
    TrebleReducer(PresetEqualizerProfile.TrebleReducer),
    Custom(null);

    fun toEqualizerConfiguration(volumeOffsets: ByteArray?): EqualizerConfiguration {
        return if (presetProfile != null) {
            EqualizerConfiguration(presetProfile)
        } else {
            if (volumeOffsets != null) {
                EqualizerConfiguration(EqualizerBandOffsets(volumeOffsets))
            } else {
                throw NullPointerException("volumeOffsets is null")
            }
        }
    }

    companion object {
        fun fromPresetProfile(presetProfile: PresetEqualizerProfile?): EqualizerProfile {
            EqualizerProfile.values().forEach {
                if (it.presetProfile == presetProfile) {
                    return it
                }
            }
            Log.e(
                "EqualizerProfile",
                "Couldn't find EqualizerProfile for preset $presetProfile, using Custom",
            )
            return Custom
        }
    }
}

File: web/src/components/EqualizerProfile.ts
import { PresetEqualizerProfile } from "../../wasm/pkg/openscq30_web_wasm";

export type PresetProfileName = (typeof presetProfiles)[number]["name"];
export const presetProfiles = [
  { name: "SoundcoreSignature", id: PresetEqualizerProfile.SoundcoreSignature },
  { name: "Acoustic", id: PresetEqualizerProfile.Acoustic },
  { name: "BassBooster", id: PresetEqualizerProfile.BassBooster },
  { name: "BassReducer", id: PresetEqualizerProfile.BassReducer },
  { name: "Classical", id: PresetEqualizerProfile.Classical },
  { name: "Podcast", id: PresetEqualizerProfile.Podcast },
  { name: "Dance", id: PresetEqualizerProfile.Dance },
  { name: "Deep", id: PresetEqualizerProfile.Deep },
  { name: "Electronic", id: PresetEqualizerProfile.Electronic },
  { name: "Flat", id: PresetEqualizerProfile.Flat },
  { name: "HipHop", id: PresetEqualizerProfile.HipHop },
  { name: "Jazz", id: PresetEqualizerProfile.Jazz },
  { name: "Latin", id: PresetEqualizerProfile.Latin },
  { name: "Lounge", id: PresetEqualizerProfile.Lounge },
  { name: "Piano", id: PresetEqualizerProfile.Piano },
  { name: "Pop", id: PresetEqualizerProfile.Pop },
  { name: "RnB", id: PresetEqualizerProfile.RnB },
  { name: "Rock", id: PresetEqualizerProfile.Rock },
  { name: "SmallSpeakers", id: PresetEqualizerProfile.SmallSpeakers },
  { name: "SpokenWord", id: PresetEqualizerProfile.SpokenWord },
  { name: "TrebleBooster", id: PresetEqualizerProfile.TrebleBooster },
  { name: "TrebleReducer", id: PresetEqualizerProfile.TrebleReducer },
] as const;
